scylla-proxy/src/proxy.rs

```diff
@@ -681,22 +683,27 @@ impl Doorkeeper {
         real_addr: SocketAddr,
         shard_awareness: ShardAwareness,
     ) -> Result<TcpStream, DoorkeeperError> {
-        let cluster_stream = if shard_awareness.is_aware() {
-            let shards = match self.shards_number {
-                None => {
-                    let temporary_stream = TcpStream::connect(real_addr)
-                        .await
-                        .map_err(|err| DoorkeeperError::NodeConnectionAttempt(real_addr, err))?;
-                    let shards = self
-                        .obtain_shards_number(temporary_stream, real_addr)
-                        .await?;
-                    self.shards_number = Some(shards);
-
-                    shards
-                }
-                Some(shards) => shards,
-            };
+        let shards = if shard_awareness.is_aware() {
+            if let shards @ Some(_) = self.shards_number {
+                shards
+            } else {
+                let temporary_stream = TcpStream::connect(real_addr)
+                    .await
+                    .map_err(|err| DoorkeeperError::NodeConnectionAttempt(real_addr, err))?;
+                let shards = match self.obtain_shards_number(temporary_stream, real_addr).await {
+                    Ok(shards) => Some(shards),
+                    // If a node offers no sharding info, change proxy ShardAwareness to Unaware.
+                    Err(DoorkeeperError::ObtainingShardNumberNoShardInfo) => None,
+                    Err(e) => return Err(e),
+                };
+                self.shards_number = shards;
```

Comment:
By the way, I think it's better to avoid the word "number" for variables which indicate a quantity of items. The `shards_number` variable could be interpreted both as "the number of shards" and "shard's number" (a.k.a. "shard's ID"). The word "count" is a better replacement as it's less ambiguous - so, e.g. `shard_count`.

Reply:
Just ranting, it's not a big issue, it's fine to fix this later.